File: 01-Languages/Go/08-Strings/main.go
// main.go - common string operations, byte vs rune, strconv conversions.
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {
	fmt.Println("--- common operations ---")
	s := "  hello  "
	fmt.Printf("TrimSpace: [%s]\n", strings.TrimSpace(s))
	fmt.Println("ToUpper:", strings.ToUpper(s))
	fmt.Println("Contains 'ell':", strings.Contains(s, "ell"))
	fmt.Println("Split:", strings.Split("a,b,c", ","))
	fmt.Println("Join:", strings.Join([]string{"a", "b"}, "-"))
	fmt.Println("ReplaceAll:", strings.ReplaceAll("hello", "l", "L"))

	fmt.Println("\n--- byte length vs rune count ---")
	accented := "héllo"
	fmt.Println("len(accented) [byte count]:", len(accented))
	runeCount := 0
	for range accented {
		runeCount++
	}
	fmt.Println("rune count (via for range):", runeCount)

	fmt.Println("\n--- rune iteration ---")
	for i, r := range accented {
		fmt.Printf("byte offset %d: %c\n", i, r)
	}

	fmt.Println("\n--- strconv ---")
	n, err := strconv.Atoi("42")
	fmt.Println("Atoi(\"42\"):", n, "err:", err)

	_, err2 := strconv.Atoi("not-a-number")
	fmt.Println("Atoi(\"not-a-number\") err:", err2)

	str := strconv.Itoa(42)
	fmt.Println("Itoa(42):", str)
}

File: 01-Languages/Rust/08-Strings/main.rs
// main.rs - String vs &str, no direct integer indexing, byte length vs char count, common ops.

fn takes_str(s: &str) -> usize {
    s.len()
}

fn main() {
    println!("--- String vs &str ---");
    let owned: String = String::from("hello");
    let borrowed: &str = "hello";
    println!("takes_str(&owned): {}", takes_str(&owned));
    println!("takes_str(borrowed): {}", takes_str(borrowed));

    println!("\n--- no direct integer indexing; use .chars() instead ---");
    let s = String::from("héllo");
    let first_char = s.chars().next();
    println!("first_char: {:?}", first_char);
    // let x = s[0]; // would fail to COMPILE: `String` cannot be indexed by `{integer}`

    println!("\n--- byte length vs char count ---");
    println!("s.len() [byte length]: {}", s.len());
    println!("s.chars().count() [actual character count]: {}", s.chars().count());

    println!("\n--- common operations ---");
    let padded = "  hello  ";
    println!("trim: [{}]", padded.trim());
    println!("to_uppercase: {}", padded.to_uppercase());
    println!("contains 'ell': {}", padded.contains("ell"));
    println!("split: {:?}", "a,b,c".split(',').collect::<Vec<&str>>());
    println!("join: {}", ["a", "b"].join("-"));
}
